daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/layers/LayerDriver.scala

```diff
@@ -0,0 +1,338 @@
+      case pe: ProcessingError if pe.maybeCause.isDefined =>
+        lr.runtimeSchemaDefinitionError(pe.maybeCause.get)
+      case pe: ProcessingError if pe.maybeFormatString.isDefined =>
+        lr.runtimeSchemaDefinitionError(pe.maybeFormatString.get)
+      case e: Exception =>
+        lr.runtimeSchemaDefinitionError(e)
+      case _ => throw t
+    }
+  }
+
+  final def addOutputLayer(s: DataOutputStream): DirectOrBufferedDataOutputStream = {
+    val jos = wrapJavaOutputStream(s)
+    val wrappedStream =
+      try {
+        layer.wrapLayerOutput(jos)
+      } catch {
+        case t: Throwable => escalateThrowFromWrap(t)
+      }
+    val encodedOutputStream = new ThrowProtectedAssuredCloseOutputStream(wrappedStream, jos)
+    val newDOS = DirectOrBufferedDataOutputStream(
+      encodedOutputStream,
+      null,
+      isLayer = true,
+      s.chunkSizeInBytes,
+      s.maxBufferSizeInBytes,
+      s.tempDirPath,
+    )
+    newDOS.setPriorBitOrder(BitOrder.MostSignificantBitFirst)
+    newDOS.setAbsStartingBitPos0b(ULong(0L))
+    newDOS.setDebugging(s.areDebugging)
+    newDOS
+  }
+
+  /**
+   * Unparsing is very asynchronous due to suspensions.
+   * So what is done here has to arrange for the layer stream to be closed,
+   * but in the future when everything that has been written to it
+   * but was suspended, has actually been written.
+   */
+  final def removeOutputLayer(s: DirectOrBufferedDataOutputStream, state: UState): Unit = {
+    // do nothing
+  }
+
+  private final class AssuredCloseInputStream(stream: InputStream, inner: InputStream)
+    extends FilterInputStream(stream) {
+    override def close(): Unit = {
+      stream.close()
+      inner.close()
+    }
+  }
+
+  /**
+   * Besides making sure the inner stream gets closed, this also
+   * provides handling of exceptions thrown by the layer-author's code
+   * (or things that that code calls).
+   *
+   * This is needed because unparsing can involve suspensions, so these
+   * calls can happen long after the call stack of unparsers is over; hence,
+   * any nested try/catches around the actual unparser calls may not be
+   * surrounding these actions.
+   *
+   * To get sensible behavior even if these throws happen late due to
+   * suspensions, we have to encapsulate here with our own throw/catches.
+   *
+   * @param stream the stream we're encapsulating
+   * @param inner the stream we're assuring will be closed even if a close on the primary stream doesn't propagate the stream
+   */
+  private final class ThrowProtectedAssuredCloseOutputStream(
+    stream: OutputStream,
+    inner: OutputStream,
+  ) extends OutputStream {
+
+    /**
+     * We hope this isn't get used because a try-catch around every individual write of a byte
+     * is inefficient certainly.
+     * @param b
+     */
+    override def write(b: Int): Unit = {
+      try {
+        stream.write(b)
+      } catch {
+        case e: Exception => handleOutputStreamException(e)
```

Comment:
Again surprising. There are definitely tests that throw Exception from within the write to a layer stream.